### ISSUE TITLE
ProvideMany: high memory usage when providing tens of millions of CIDs

### ISSUE THREAD

**Issue by ajnavarro** [ipfs/go-libipfs#354]
When using BatchProviding, we are not really batching, but sending all the CIDs at the same time to the Router implementing ProvideMany.

To avoid collateral problems, we should actually batch the calls to ProvideMany.

This will help with Reframe Router implementation (https://github.com/ipfs/go-delegated-routing) to avoid huge JSON payloads sent to the server.

We need to find good defaults to still keep FullRT DHT implementation with good performance numbers.

That is like 1/10th of the memory spike observed, we are still searching for other possible problems.

**Comment by ajnavarro** [comment]
CC  @ischasny

**Comment by BigLep** [comment]
2022-10-18 conversation: we aren't aware of this being a blocker at the moment so not prioritizing currently, but feedback welcome if this needs to be moved up sooner.

**Comment by ischasny** [comment]
@ajnavarro as per our discussion, might be a good idea to chunk up the CIDs snapshot into smaller pieces so that at least we don't squeeze all of them into a single HTTP request. That is problematic on larger nodes (like web3 storage). Snapshots don't get reprovided even with high router timeouts. Maybe we can do that only for reframe router initially? That should save us some memory on both sending and receiving side. Wdyt?

**Comment by ajnavarro** [comment]
Yeah, won't be the final solution, but will help in providing over HTTP.

**Comment by ischasny** [comment]
Great! Would you guys be up for taking it into the next release? Should be simple to do and would unblock us too.

**Comment by ajnavarro** [comment]
Related issue: https://github.com/ipfs/go-delegated-routing/issues/55
